common/dbutils/changeset.go

```diff
@@ -182,56 +132,171 @@ func (s *ChangeSet) Add(key []byte, value []byte) error {
+	return int(binary.BigEndian.Uint32(b[4:8]))
+}
+
+func checkKeySize(b []byte, key []byte) error {
+	size := KeySize(b)
+
+	if Len(b) == 0 || len(key) == size {
+		return nil
+	}
+
+	return fmt.Errorf("wrong key size in ChangeSet: expected %d, actual %d", size, len(key))
+}
+
+// Add adds a new entry to the ChangeSet.
+// One must not add an existing key
 // and may add keys only of the same size.
-func (s *ChangeSet) MultiAdd(changes []Change) error {
-	for i := range changes {
-		if err := s.checkKeySize(changes[i].Key); err != nil {
-			return err
+func Add(b []byte, key []byte, value []byte) ([]byte, error) {
+	var m int
+	var n int
+
+	if len(b) == 0 {
+		m = len(key)
+		n = 0
+	} else {
+		m = KeySize(b)
+		n = Len(b)
+		if err := checkKeySize(b, key); err != nil {
+			return nil, err
 		}
-		s.Changes = append(s.Changes, changes[i])
+	}
+	pos := 4
+	var buffer bytes.Buffer
+	// Encode n
+	intArr := make([]byte, 4)
+	binary.BigEndian.PutUint32(intArr, uint32(n+1))
+	buffer.Write(intArr)
+	// KeySize should be the same
+	if n == 0 {
+		binary.BigEndian.PutUint32(intArr, uint32(len(key)))
+		buffer.Write(intArr)
+	} else {
+		buffer.Write(b[pos : pos+4])
 	}
 
-	return nil
+	pos += 4
+	// append key
+	if n == 0 {
+		buffer.Write(key)
+		pos += len(key)
+	} else {
+		buffer.Write(b[pos : pos+n*m])
+		buffer.Write(key)
+		pos += n * m
+	}
+	// Append Index
+	if n == 0 {
+		binary.BigEndian.PutUint32(intArr, uint32(len(value)))
+		buffer.Write(intArr)
+	} else {
+		buffer.Write(b[pos : pos+4*n])
+		pos += 4 * n
+		prev := int(binary.BigEndian.Uint32(b[pos-4 : pos]))
+		binary.BigEndian.PutUint32(intArr, uint32(prev+len(value)))
+		buffer.Write(intArr)
+		buffer.Write(b[pos:])
+	}
+	// Append Value
+	buffer.Write(value)
+	return buffer.Bytes(), nil
 }
 
-func (s *ChangeSet) Walk(f func(k, v []byte) error) error {
-	for i := range s.Changes {
-		err := f(s.Changes[i].Key, s.Changes[i].Value)
+func Walk(b []byte, f func(k, v []byte) error) error {
+	if len(b) == 0 {
+		return nil
+	}
+	if len(b) < 8 {
+		return fmt.Errorf("decode: input too short (%d bytes)", len(b))
+	}
+
+	n := binary.BigEndian.Uint32(b[0:4])
+	m := binary.BigEndian.Uint32(b[4:8])
+
+	if n == 0 {
+		return nil
+	}
+	valOffset := 8 + n*m + 4*n
+	if uint32(len(b)) < valOffset {
+		return fmt.Errorf("decode: input too short (%d bytes, expected at least %d bytes)", len(b), valOffset)
+	}
+
+	totalValLength := binary.BigEndian.Uint32(b[valOffset-4 : valOffset])
+	if uint32(len(b)) < valOffset+totalValLength {
+		return fmt.Errorf("decode: input too short (%d bytes, expected at least %d bytes)", len(b), valOffset+totalValLength)
+	}
+
+	for i := uint32(0); i < n; i++ {
+		key := b[8+i*m : 8+(i+1)*m]
+		idx0 := uint32(0)
+		if i > 0 {
+			idx0 = binary.BigEndian.Uint32(b[8+n*m+4*(i-1) : 8+n*m+4*i])
+		}
+		idx1 := binary.BigEndian.Uint32(b[8+n*m+4*i : 8+n*m+4*(i+1)])
+		val := b[valOffset+idx0 : valOffset+idx1]
+
+		err := f(common.CopyBytes(key), common.CopyBytes(val))
```

Comment:
It should be ok not to make copies of key and val, and document it in the comment on top of the function. We can check that in all existing cases the copy is not required